src/plasmapy/simulation/particle_tracker/particle_tracker.py

```diff
@@ -620,117 +703,205 @@ def particles_on_grid(self):
+        )
+        dx = np.multiply(current_speeds, self.dt)
+
+        stopping_power = np.zeros((self.nparticles_tracked, 1))
+        relevant_kinetic_energy = (
+            self._particle_kinetic_energy[self._tracked_particle_mask] * u.J
+        )
+
+        for cs in self._stopping_power_interpolators:
+            interpolation_result = (
+                cs(relevant_kinetic_energy).to(u.J * u.m**2 / u.kg).value
+            )
+
+            stopping_power += interpolation_result
+
+        # Take the negative stopping power since we want to be removing energy
+        # from the particles
+
+        energy_loss_per_length = np.multiply(
+            stopping_power,
+            summed_field_values["rho"].to(u.kg / u.m**3).value[:, np.newaxis],
+        )
+
+        dE = -np.multiply(energy_loss_per_length, dx)
+
+        # Update the velocities of the particles using the new energy values
+        # TODO: again, figure out how to differentiate relativistic and classical cases
+        E = self._particle_kinetic_energy[self._tracked_particle_mask] + dE
+
+        particles_to_be_stopped_mask = np.full(
+            shape=self._tracked_particle_mask.shape, fill_value=False
+        )
+        tracked_particles_to_be_stopped_mask = (
+            E < 0
+        ).flatten()  # A subset of the tracked particles!
+        # Of the tracked particles, stop the ones indicated by the subset mask
+        particles_to_be_stopped_mask[self._tracked_particle_mask] = (
+            tracked_particles_to_be_stopped_mask
         )
 
-        self.dt = dt
+        new_speeds = np.sqrt(2 * E / self.m)
+        self.v[self._tracked_particle_mask] = np.multiply(
+            new_speeds, velocity_unit_vectors
+        )
+
+        self._stop_particles(particles_to_be_stopped_mask)
+
+    def _push(self) -> None:
+        r"""
+        Advance particles using an implementation of the time-centered
+        Boris algorithm.
+        """
+
+        # Interpolate fields at particle positions
+        summed_field_values, E, B = self._interpolate_grid()
+
+        # Calculate an appropriate timestep (uniform, synchronized)
+        self.dt = self._update_time(summed_field_values)
+
+        # Update the position and velocities of the particles using timestep
+        # calculations as well as the magnitude of E and B fields
+        self._update_position(B, E)
+
+        if not self._integrator.is_relativistic and not self._raised_relativity_warning:
+            beta_max = self.vmax / const.c.si.value
+
+            if beta_max >= 0.01:
+                self._raised_relativity_warning = True
+
+                warnings.warn(
+                    f"Particles have reached {beta_max}% of the speed of light. Consider using a relativistic integrator for more accurate results.",
+                    RelativityWarning,
+                )
+
+        # Update velocities to reflect stopping
+        if not self._do_stopping:
```

Comment:
We are going to add more to this function later - you don't want it to stop here if you are doing scattering but not stopping. Why not do 

```
if stopping: 
   self._update_velocity_stopping
```
?